File: 3-1.py
#!/usr/bin/env python

from functools import reduce


def priority(item):
    if ord(item) in range(ord('a'), ord('z') + 1):
        return ord(item) - 96
    if ord(item) in range(ord('A'), ord('Z') + 1):
        return ord(item) - 64 + 26


rucksacks = [row.strip() for row in open('3.input').readlines()]
rucksacks = [(r[:len(r)//2], r[len(r)//2:]) for r in rucksacks]
duplicates = [set(r[0]) & set(r[1]) for r in rucksacks]
print(reduce(lambda x, y: x + y, [priority(d.pop()) for d in duplicates]))

File: 3-1.rb
#!/usr/bin/env ruby

require 'set'

def priority(item)
  case item
  when 'a'..'z'
    item.ord - 96
  when 'A'..'Z'
    item.ord - 64 + 26
  end
end

print File.read('3.input').lines.map(&:strip).
        map { |r| [r[.. r.length / 2 - 1], r[r.length / 2 ..]] }.
        map { |r| Set.new(r[0].chars) & Set.new(r[1].chars) }.
        map(&:to_a).
        map { |duplicate| priority(duplicate[0]) }.
        reduce(:+)
print "\n"
